### PR TITLE
[PT FE] Enable reverse infer

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *No need to specify input type and rank for majority of models*

### Tickets:
 - *102349*

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 1]
```diff
@@ -111,6 +137,95 @@ bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::
...
+                }
+                if (if_op->get_input_tensor(in_indx).get_element_type().is_dynamic()) {
+                    if_op->get_input_tensor(in_indx).m_element_type =
+                        else_body_params.at(body_indx)->get_element_type();
+                    is_changed = true;
+                }
+            }
+        } else if (std::dynamic_pointer_cast<ConvertLike>(op)) {
+            is_changed |= inherit_output_shape(op, {0});
+            is_changed |= inherit_output_type(op, {1});
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:228]: please add tests

**Comment by rkazants** [comment]
@mvafin, please add unit-tests for the added operations

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 2]
```diff
@@ -111,6 +137,95 @@ bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::
                 }
             }
             is_changed |= inherit_output_type(op, {0, 1});
+        } else if (std::dynamic_pointer_cast<Concat>(op)) {
+            std::vector<size_t> inp_idxs(op->get_input_size());
+            std::iota(inp_idxs.begin(), inp_idxs.end(), 0);
+            is_changed |= inherit_output_rank(op, inp_idxs);
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:143]: to be more precise, you can also deduce dimension sizes except one along which inputs are concatenated. Mya be to add todo?

File: src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp [hunk 3]
```diff
@@ -111,6 +137,95 @@ bool ov::pass::ReverseShapeAndTypeInfer::run_on_model(const std::shared_ptr<ov::
...
+            is_changed |= inherit_output_rank(op, {0});
+            is_changed |= inherit_output_type(op, {0});
+        } else if (std::dynamic_pointer_cast<Squeeze>(op)) {
+            auto in0_rank = op->get_input_partial_shape(0).rank();
+            if (output_shape.rank().is_static() && in0_rank.is_dynamic() && op->get_input_size() > 1) {
+                auto in1_pshape = op->get_input_partial_shape(1);
+                if (in1_pshape.is_static()) {
+                    auto num_dims = in1_pshape.size() == 0 ? 1 : in1_pshape[0].get_length();
+                    op->get_input_tensor(0).m_partial_shape =
+                        PartialShape::dynamic(output_shape.rank().get_length() - num_dims);
```
Comment (rkazants) [inline src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp:155]: what if we have squeezing to a scalar from 1D tensor with one element, we will get `output_shape.rank().get_length() - num_dims = 0`

Comment (mvafin) [reply src/common/transformations/src/transformations/common_optimizations/reverse_shape_and_type_infer.cpp]: Fixed

**Review by rkazants — APPROVED** [review]

**Pull request merged**